scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -338,12 +369,123 @@ def highly_variable_genes(
-            arrays.extend([
-                df['highly_variable_nbatches'].values,
-                df['highly_variable_intersection'].values,
-            ])
-            dtypes.append([
-                ('highly_variable_nbatches', int),
-                ('highly_variable_intersection', np.bool_),
-            ])
+            arrays.extend(
+                [
+                    df['highly_variable_nbatches'].values,
+                    df['highly_variable_intersection'].values,
+                ]
+            )
+            dtypes.append(
+                [
+                    ('highly_variable_nbatches', int),
+                    ('highly_variable_intersection', np.bool_),
+                ]
+            )
         return np.rec.fromarrays(arrays, dtype=dtypes)
+
+
+def highly_variable_genes_seurat_v3(
+    adata: AnnData,
+    n_top_genes: int = 2000,
+    batch_key: Optional[str] = None,
+    lowess_frac: Optional[float] = 0.15,
+):
+    """\
+    Annotate highly variable genes [Stuart19]_.
+
+    Expects raw count data.
+
+    The major difference in this implementation is the use of lowess insted of loess.
+
+    For further details of the sparse arithmetic see https://www.overleaf.com/read/ckptrbgzzzpg
+    Parameters
+    ----------
+    adata
+        The annotated data matrix of shape `n_obs` × `n_vars`. Rows correspond
+        to cells and columns to genes.
+    n_top_genes
+        Number of highly-variable genes to keep.
+    batch_key
+        If specified, highly-variable genes are selected within each batch separately and merged.
+        This simple process avoids the selection of batch-specific genes and acts as a
+        lightweight batch correction method.
+    lowess_frac
+        The fraction of the data (cells) used when estimating the variance in the lowess model fit.
+    """
+    import statsmodels
+
+    lowess = statsmodels.nonparametric.lowess
+
+    if batch_key is None:
+        batch_info = pd.Categorical(np.zeros((adata.X.shape[0])).astype(int))
+    else:
+        batch_info = adata.obs[batch_key]
+
+    norm_gene_vars = []
+    for b in np.unique(batch_info):
+
+        mean, var = materialize_as_ndarray(_get_mean_var(adata[batch_info == b].X))
+        not_const = var > 0
+        estimat_var = np.zeros((adata.X.shape[1]))
+
+        y = np.log10(var[not_const])
+        x = np.log10(mean[not_const])
+        # output is sorted by x
+        v = lowess(y, x, frac=lowess_frac)
+        estimat_var[not_const][np.argsort(x)] = v[:, 1]
+
+        # get normalized variance
+        reg_std = np.sqrt(10 ** estimat_var)
+        batch_counts = adata[batch_info == b].X.copy()
+        # clip large values as in Seurat
+        N = np.sum(batch_info == b)
+        vmax = np.sqrt(N)
+        clip_val = reg_std * vmax + mean
+        # could be something faster here
+        for g in range(batch_counts.shape[1]):
+            batch_counts[:, g][batch_counts[:, g] > vmax] = clip_val[g]
```

Comment:
```suggestion
        if sp_sparse.issparse(batch_counts):
            batch_counts = sp_sparse.csr_matrix(batch_counts)
            mask = batch_counts.data > vmax
            batch_counts.data[mask] = clip_val[batch_counts.indices[mask]]
        else:
            for g in range(batch_counts.shape[1]):
                batch_counts[:, g][batch_counts[:, g] > vmax] = clip_val[g]
```

I believe this should work. Could probably be optimized further by using a `csc` matrix, and not iterating over all the genes. This version takes makes the whole call take about half a second for 2700 cells x 16634 genes in csr format, and 1.15 seconds for a dense array.

Reply:
good catch, this looks like a nice solution